### PR TITLE
Add VTT style properties and regions

### PR DESCRIPTION
**Pull request by vzablock**
Added styling of new VTT properties and regions.

VTT subtitles now have supported styling that comes from the `.vtt` file.
Reference:
 - https://www.w3.org/TR/webvtt1/
 - https://www.speechpad.com/captions/webvtt

### CODE REVIEW

File: src/ts/components/subtitleoverlay.ts [hunk 1]
```diff
@@ -67,9 +69,14 @@ export class SubtitleOverlay extends Container<ContainerConfig> {
         this.subtitleContainerManager.removeLabel(this.previewSubtitle);
       }
 
-      this.subtitleContainerManager.addLabel(labelToAdd);
-      this.updateComponents();
       this.show();
```
Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]: Why moving this after the show now?

File: src/ts/components/subtitleoverlay.ts [hunk 2]
```diff
@@ -453,8 +466,8 @@ export class SubtitleRegionContainerManager {
    * If the subtitle has positioning information it is added to the container.
    * @param label The subtitle label to wrap
    */
-  addLabel(label: SubtitleLabel): void {
-    const regionName = label.region || 'default';
+  addLabel(label: SubtitleLabel, overlaySize?: {width: number, height: number}): void {
```
Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]:
```suggestion
  addLabel(label: SubtitleLabel, overlaySize?: { width: number, height: number }): void {
```

File: src/ts/components/subtitleoverlay.ts (hunk 1 above)
Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]: Why moving after show

File: src/ts/components/subtitleoverlay.ts [hunk 3]
```diff
@@ -453,8 +466,8 @@ export class SubtitleRegionContainerManager {
...
+    const regionName = label.vtt && label.vtt.region ? label.vtt.region.id : label.region || 'default';
```
Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]: Would it make sense to add brackets here? How is this `||` interpreted if e.g. `region.id` would be undefined?

File: src/ts/components/subtitleoverlay.ts [hunk 4]
```diff
@@ -474,22 +487,20 @@ export class SubtitleRegionContainerManager {
...
-    this.subtitleRegionContainers[regionName].addLabel(label);
+    this.subtitleRegionContainers[regionName].addLabel(label, overlaySize);
   }
 
   /**
    * Removes a subtitle label from a container.
    */
   removeLabel(label: SubtitleLabel): void {
-    const region = label.region || 'default';
-    for (const regionName in this.subtitleRegionContainers) {
```
Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]: do you know why this loop was there before?

File: src/ts/components/subtitleoverlay.ts [hunk 5]
```diff
@@ -516,19 +528,80 @@ class SubtitleRegionContainer extends Container<ContainerConfig> {
...
-  addLabel(labelToAdd: SubtitleLabel) {
+  private addCueBoxContainer(labelToAdd: SubtitleLabel) {
+    const id = labelToAdd.getConfig().id;
+    this.cueBoxContainers[id] = new SubtitleCueBoxContainer();
+    this.cueBoxContainers[id].addLabel(labelToAdd);
+    setVttCueBoxStyles(this.cueBoxContainers[id], labelToAdd.vtt);
+    this.addComponent(this.cueBoxContainers[id]);
+    this.updateComponents();
+
+    if (labelToAdd.vtt.region && labelToAdd.vtt.region.scroll) {
```
Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]: for backwards compatiliby also check if `labelToAdd.vtt` is not empty.

Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]: What does scroll mean exactly? (There is no documentation in the player)

File: src/ts/components/subtitleoverlay.ts [hunk 6]
```diff
@@ -505,8 +516,9 @@ export class SubtitleRegionContainerManager {
   }
 }
 
-class SubtitleRegionContainer extends Container<ContainerConfig> {
+export class SubtitleRegionContainer extends Container<ContainerConfig> {
   private labelCount = 0;
```
Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]: I guess the `labelCount` can be removed now

File: src/ts/components/subtitleoverlay.ts [hunk 7]
```diff
@@ -516,19 +528,80 @@ class SubtitleRegionContainer extends Container<ContainerConfig> {
     }, this.config);
   }
 
-  addLabel(labelToAdd: SubtitleLabel) {
+  private addCueBoxContainer(labelToAdd: SubtitleLabel) {
+    const id = labelToAdd.getConfig().id;
+    this.cueBoxContainers[id] = new SubtitleCueBoxContainer();
```
Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]: Is there always just a 1-to-1 relationship as you always create a new container?

File: src/ts/components/subtitleoverlay.ts [hunk 8]
```diff
@@ -516,19 +528,80 @@ class SubtitleRegionContainer extends Container<ContainerConfig> {
...
   }
 
   public isEmpty(): boolean {
     return this.labelCount === 0;
   }
-}
\ No newline at end of file
+}
+
+export class SubtitleCueBoxContainer extends Container<ContainerConfig> {
```
Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]: why is this needed? I think the idea of the `SubtitleRegionContainer` is exactly that.

File: src/ts/components/subtitlevtt.ts [hunk 9]
```diff
@@ -0,0 +1,142 @@
+import {SubtitleRegionContainer, SubtitleCueBoxContainer} from './subtitleoverlay';
+import { VTTProperties, VTTRegionProperties } from 'bitmovin-player/types/subtitles/vtt/API';
+
+type Direction = 'top' | 'bottom' | 'left' | 'right';
+
+const lineHeight = 28;
```
Comment (stonko1994) [inline src/ts/components/subtitlevtt.ts]: where are those numbers coming from? Maybe add a reference to the vtt-spec.

File: src/ts/components/subtitlevtt.ts [hunk 10]
```diff
@@ -0,0 +1,142 @@
+import {SubtitleRegionContainer, SubtitleCueBoxContainer} from './subtitleoverlay';
```
Comment (stonko1994) [inline src/ts/components/subtitlevtt.ts]:
```suggestion
import { SubtitleRegionContainer, SubtitleCueBoxContainer } from './subtitleoverlay';
```

File: src/ts/components/subtitlevtt.ts [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+import { VTTProperties, VTTRegionProperties } from 'bitmovin-player/types/subtitles/vtt/API';
+
+type Direction = 'top' | 'bottom' | 'left' | 'right';
```
Comment (stonko1994) [inline src/ts/components/subtitlevtt.ts]: Why not using an enum here?

File: src/ts/components/subtitlevtt.ts [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    cueContainer.getDomElement().css('right', 'unset');
+    setVttPositionAlign(cueContainer, vtt, 'left');
+  } else if (vtt.position !== 'auto' && vtt.vertical != null) {
+    cueContainer.getDomElement().css('height', `${containerSize}%`);
+    setVttPositionAlign(cueContainer, vtt, 'top');
+  }
+};
+
+// https://www.w3.org/TR/webvtt1/#regions
+export const setVttRegionStyles = (regionContainer: SubtitleRegionContainer | SubtitleRegionContainer, region: VTTRegionProperties, overlaySize: {width: number, height: number}) => {
```
Comment (stonko1994) [inline src/ts/components/subtitlevtt.ts]: duplicate declaration (same type)

File: src/ts/components/subtitlevtt.ts (hunk 9 above)
Comment (stonko1994) [inline src/ts/components/subtitlevtt.ts]: How does this play together with our subtitle customization?

File: src/ts/components/subtitlevtt.ts [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+  setVttWritingDirection(cueContainer, vtt);
+
+  // https://w3.org/TR/webvtt1/#webvtt-cue-text-alignment
+  const textAlign = vtt.align === 'middle' ? 'center' : vtt.align;
+  cueContainer.getDomElement().css('text-align', textAlign);
+
+  // https://w3.org/TR/webvtt1/#webvtt-cue-size
+  const containerSize = vtt.size;
+  if (vtt.position !== 'auto' && vtt.vertical == null) {
```
Comment (stonko1994) [inline src/ts/components/subtitlevtt.ts]: According to the player interface `vtt.vertical` is non-optional

File: src/ts/components/subtitleoverlay.ts (hunk 1 above)
Comment (vzablock) [reply src/ts/components/subtitleoverlay.ts]: In order to get the relative sizes and positioning calculations of the `VTTRegion` we have to show the element in the DOM to get those relative values as pixels for calculations

File: src/ts/components/subtitleoverlay.ts (hunk 4 above)
Comment (vzablock) [reply src/ts/components/subtitleoverlay.ts]: Not sure but @hanneslinder said that is is fine for us to remove them as they do not affect anything

File: src/ts/components/subtitleoverlay.ts (hunk 5 above)
Comment (vzablock) [reply src/ts/components/subtitleoverlay.ts]: Means that the bottom text pushes the top one (in our css means that we scroll the subtitle container to the bottom) and hide the subtitles on the top

File: src/ts/components/subtitleoverlay.ts (hunk 6 above)
Comment (vzablock) [reply src/ts/components/subtitleoverlay.ts]: There is a public method using the count to remove the region container itself if there are no more labels inside

File: CHANGELOG.md [hunk 14]
```diff
@@ -6,6 +6,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [develop]
 
+### Added
+- Subtitle stlying with VTT properties and regions
```
Comment (stonko1994) [inline CHANGELOG.md]:
```suggestion
- Support for regions in VTT subtitles
```

To be consistent with the TTML entry

File: src/ts/components/subtitleoverlay.ts [hunk 15]
```diff
@@ -453,43 +466,65 @@ export class SubtitleRegionContainerManager {
...
+        cssClasses
       });
 
-      this.subtitleRegionContainers[regionName] = regionContainer;
+      this.subtitleRegionContainers[regionContainerId] = regionContainer;
 
       if (label.regionStyle) {
         regionContainer.getDomElement().attr('style', label.regionStyle);
+      } else if (label.vtt && !label.vtt.region) {
+        regionContainer.getDomElement().css('position', 'unset');
```
Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]: Maybe add a short comment why this is needed.

File: src/ts/components/subtitleoverlay.ts [hunk 16]
```diff
@@ -516,8 +551,18 @@ class SubtitleRegionContainer extends Container<ContainerConfig> {
     }, this.config);
   }
 
-  addLabel(labelToAdd: SubtitleLabel) {
+  addLabel(labelToAdd: SubtitleLabel, overlaySize?: {width: number, height: number}) {
     this.labelCount++;
+
+    if (labelToAdd.vtt) {
+      debugger;
```
Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]: Please remove debugger statement.

File: src/ts/components/subtitleoverlay.ts [hunk 17]
```diff
@@ -516,8 +551,18 @@ class SubtitleRegionContainer extends Container<ContainerConfig> {
     }, this.config);
   }
 
-  addLabel(labelToAdd: SubtitleLabel) {
+  addLabel(labelToAdd: SubtitleLabel, overlaySize?: {width: number, height: number}) {
```
Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]:
```suggestion
  addLabel(labelToAdd: SubtitleLabel, overlaySize?: { width: number, height: number }) {
```

File: src/ts/components/subtitleoverlay.ts [hunk 18]
```diff
@@ -453,43 +466,65 @@ export class SubtitleRegionContainerManager {
    * If the subtitle has positioning information it is added to the container.
    * @param label The subtitle label to wrap
    */
-  addLabel(label: SubtitleLabel): void {
-    const regionName = label.region || 'default';
-    if (!this.subtitleRegionContainers[regionName]) {
+  addLabel(label: SubtitleLabel, overlaySize?: {width: number, height: number}): void {
```
Comment (stonko1994) [inline src/ts/components/subtitleoverlay.ts]:
```suggestion
  addLabel(label: SubtitleLabel, overlaySize?: { width: number, height: number }): void {
```

File: src/ts/components/subtitlevtt.ts [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+  [Direction.Top, Direction.Bottom],
+  [Direction.Left, Direction.Right],
+  [Direction.Right, Direction.Left],
+]);
+
+/**
+ * Sets the default standardized styles for the Cue Box
+ * https://w3.org/TR/webvtt1/#applying-css-properties
+ */
+const setDefaultVttStyles = (cueContainer: SubtitleLabel) => {
```
Comment (stonko1994) [inline src/ts/components/subtitlevtt.ts]: I think we usually add such methods into a namespace.

File: src/ts/components/subtitlevtt.ts [hunk 20]
```diff
@@ -0,0 +1,157 @@
+import { SubtitleRegionContainer, SubtitleLabel } from './subtitleoverlay';
```
Comment (stonko1994) [inline src/ts/components/subtitlevtt.ts]: As this is no component please move one folder level up. (Or think about introducing a helper folder)

File: src/ts/components/subtitlevtt.ts [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+    cueContainer.getDomElement().css('width', `${containerSize}%`);
+    setVttPositionAlign(cueContainer, vtt, Direction.Left);
+  } else {
+    cueContainer.getDomElement().css('height', `${containerSize}%`);
+    setVttPositionAlign(cueContainer, vtt, Direction.Top);
+  }
+};
+
+// https://www.w3.org/TR/webvtt1/#regions
+export const setVttRegionStyles = (regionContainer: SubtitleRegionContainer, region: VTTRegionProperties, overlaySize: {width: number, height: number}) => {
```
Comment (stonko1994) [inline src/ts/components/subtitlevtt.ts]:
```suggestion
export const setVttRegionStyles = (regionContainer: SubtitleRegionContainer, region: VTTRegionProperties, overlaySize: { width: number, height: number }) => {
```

File: src/ts/components/subtitlevtt.ts [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+      cueContainer.getDomElement().css(DirectionPair.get(direction), `${vtt.position}%`);
+      break;
+    default:
+      cueContainer.getDomElement().css(direction, `${vtt.position}%`);
+      cueContainer.getDomElement().css(DirectionPair.get(direction), 'auto');
+  }
+};
+
+export const setVttCueBoxStyles = (cueContainer: SubtitleLabel, vtt: VTTProperties) => {
+  console.warn(vtt);
```
Comment (stonko1994) [inline src/ts/components/subtitlevtt.ts]: Please remove console output

**Comment by stonko1994** [comment]
@vzablock please also request a second reviewer.

File: src/ts/vttutils.ts [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+  [Direction.Right, Direction.Left],
+]);
+
+/**
+ * Sets the default standardized styles for the Cue Box
+ * https://w3.org/TR/webvtt1/#applying-css-properties
+ */
+const setDefaultVttStyles = (cueContainer: SubtitleLabel) => {
+  if (cueContainer.vtt.region) {
+    cueContainer.getDomElement().css('position', 'relative');
```
Comment (hanneslinder) [inline src/ts/vttutils.ts]: I think it would be better to pass the actual DOM Element instead of the `cueContainer` to all the helper functions in `vttutils` as all you ever need from it is `cueContainer.getDomElement()`.

Comment (vzablock) [reply src/ts/vttutils.ts]: The problem is that the type gets lost in that case, the parameter required then would only be `DOM` and not `SubtitleLabel`

File: spec/vttutils.spec.ts [hunk 24]
```diff
@@ -0,0 +1,572 @@
...
+
+describe('Vtt Utils', () => {
+  describe('Vtt Region', () => {
+    it('should set region css properties', () => {
+        const mockRegionContainer = generateSubtitleRegionContainerMock();
+        const mockDomElement = MockHelper.generateDOMMock();
+        mockRegionContainer.getDomElement = () => mockDomElement;
+        const spyCss = jest.spyOn(mockRegionContainer.getDomElement(), 'css');
+    
+        VttUtils.setVttRegionStyles(mockRegionContainer, vttRegionProps, {width: 1000, height: 800});
```
Comment (stonko1994) [inline spec/vttutils.spec.ts]:
```suggestion
        VttUtils.setVttRegionStyles(mockRegionContainer, vttRegionProps, { width: 1000, height: 800 });
```

File: spec/vttutils.spec.ts [hunk 25]
```diff
@@ -0,0 +1,572 @@
...
+      const spyCss = jest.spyOn(mockRegionContainer.getDomElement(), 'css');
+
+      VttUtils.setVttCueBoxStyles(mockRegionContainer);
+      
+      expect(spyCss).toHaveBeenCalledTimes(11);
+      expect(spyCss).toHaveBeenNthCalledWith(8, 'text-align', 'left');
+    });
+
+    it('should not set default position', () => {
+      const mockRegionContainer = generateSubtitleCueBoxMock(false, { position: 50 } as VTTProperties);
```
Comment (stonko1994) [inline spec/vttutils.spec.ts]: Shouldn't that be `auto` now instead of `50`?

Comment (vzablock) [reply spec/vttutils.spec.ts]: The test says "it should NOT set default" 😄
